hr_holidays_public/models/resource_calendar.py

```diff
@@ -15,77 +11,42 @@
 class ResourceCalendar(models.Model):
     _inherit = "resource.calendar"
 
-    def _public_holidays_leave_intervals(self, start_dt, end_dt, employee_id, tz):
-        """Get the public holidays for the current employee and given dates in
-        the format expected by resource methods.
-
-        :param: start_dt: Initial datetime.
-        :param: end_dt: End datetime.
-        :param: employee_id: Employee ID. It can be false.
-        :return: List of tuples with (start_date, end_date) as elements.
-        """
+    def _attendance_intervals_batch_exclude_public_holidays(
+        self, start_dt, end_dt, intervals, resources, tz
+    ):
         HrHolidaysPublic = self.env["hr.holidays.public"]
-
-        leaves = []
-        if start_dt.year != end_dt.year:
-            # This fixes the case of leave request asked over 2 years.
-            #
-            # adding 1 year to end_dt for rrule to retrieve correct years for
-            # public holidays to work
-            # rrule.rrule(rrule.YEARLY, dtstart=2019-12-22, until=2020-01-05)
-            # gives [2019]
-            # rrule.rrule(rrule.YEARLY, dtstart=2019-12-22, until=2021-01-05)
-            # gives [2019, 2020]
-            end_dt = end_dt.replace(year=end_dt.year + 1)
-
-        for day in rrule.rrule(rrule.YEARLY, dtstart=start_dt, until=end_dt):
-            lines = HrHolidaysPublic.get_holidays_list(
-                day.year, employee_id=employee_id
-            )
-            # In some cases, an error appears about mixing 2 models
-            # (hr.holidays.public.line + resource.calendar.leaves)
-            # in _leave_intervals or _leave_intervals_batch functions
-            # It only happen when both holidays and leaves exist.
-            # The solution is to pass an empty leave in the tuple instead
-            # of the public holiday line record, as this element has no
-            # further use except the union operation.
-            resource_leave_model = self.env["resource.calendar.leaves"]
-            for line in lines:
-                leaves.append(
-                    (
-                        datetime.combine(line.date, time.min).replace(tzinfo=tz),
-                        datetime.combine(line.date, time.max).replace(tzinfo=tz),
-                        resource_leave_model,
+        list_by_years = {}
+        for resource in resources:
+            interval_resource = intervals[resource.id]
+            attendances = []
+            for attendance in interval_resource._items:
+                year_item = attendance[0].date().year
+                if year_item not in list_by_years:
+                    list_by_years[year_item] = HrHolidaysPublic.get_holidays_list(
```

Comment:
You are still calling too much this method (one per resource attendance). As previously said, it's not a bad idea to have a method to retrieve by date range the public holidays. My only comment was not to change the API of the current method, creating other method in a way that reuse the same code (for example leaving the other method signature and calling a new one with dates having from 01/01 to 31/12), or putting all arguments (year, date_start and date_end) in the same method as keyword ones.